File: exercism/typescript/protein-translation/protein-translation.ts
export function translate(rnaSequence: string): string[] {
  let rnaTranslation: Map<string, string> =
    new Map([
      ["AUG", "Methionine"],
      ["UUU", "Phenylalanine"],
      ["UUC", "Phenylalanine"],
      ["UUA", "Leucine"],
      ["UUG", "Leucine"],
      ["UCU", "Serine"],
      ["UCC", "Serine"],
      ["UCA", "Serine"],
      ["UCG", "Serine"],
      ["UAU", "Tyrosine"],
      ["UAC", "Tyrosine"],
      ["UGU", "Cysteine"],
      ["UGC", "Cysteine"],
      ["UGG", "Tryptophan"],
      ["UAA", "STOP"],
      ["UAG", "STOP"],
      ["UGA", "STOP"]
    ]);

  let proteins: string[] = [];
  let codonsBits: string[] = rnaSequence.split("");
  let codon: string = ""
  for (let codonBit of codonsBits) {
    if (codon.length < 3) {
      codon += codonBit;
    } else {
      codon = codonBit
    }
    if (rnaTranslation.has(codon)) {
      if (rnaTranslation.get(codon) === "STOP") {
        break;
      } else {
        proteins.push(rnaTranslation.get(codon) ?? "");
      }
    }
  }

  return proteins;
}

File: exercism/python/protein-translation/protein_translation.py
def proteins(strand: str) -> [str]:
    mapping = {
        "AUG": "Methionine",
        "UUU": "Phenylalanine",
        "UUC": "Phenylalanine",
        "UUA": "Leucine",
        "UUG": "Leucine",
        "UCU": "Serine",
        "UCC": "Serine",
        "UCA": "Serine",
        "UCG": "Serine",
        "UAU": "Tyrosine",
        "UAC": "Tyrosine",
        "UGU": "Cysteine",
        "UGC": "Cysteine",
        "UGG": "Tryptophan",
        "UAA": "STOP",
        "UAG": "STOP",
        "UGA": "STOP"
    }

    codons: [str] = [strand[i:i + 3] for i in range(0, len(strand), 3)]
    proteins: [str] = []
    for codon in codons:
        protein: str = mapping[codon]
        if protein != "STOP":
            proteins.append(protein)
        else:
            break

    return proteins
